es-app/src/VirtualSystemManager.cpp

```diff
@@ -0,0 +1,415 @@
+}
+
+FileData::SortType VirtualSystemManager::getSortType(std::string desc) {
+	std::vector<FileData::SortType> SortTypes = FileSorts::SortTypes;
+	// find it
+	for(unsigned int i = 0; i < FileSorts::SortTypes.size(); i++)
+	{
+		const FileData::SortType& sort = FileSorts::SortTypes.at(i);
+		if(sort.description == desc)
+		{
+			return sort;
+		}
+	}
+	// if not found default to name, ascending
+	return FileSorts::SortTypes.at(0);
+}
+
+std::vector<std::string> VirtualSystemManager::stringToSystems(std::string systemString)
+{
+	// from a comma separated string, get a vector of system names
+	std::vector<std::string> strs;
+	boost::split(strs, systemString, boost::is_any_of(","));
+	return strs;
+}
+
+void VirtualSystemManager::loadVirtualSystems()
+{
+	loadAutoVirtualSystems();
+	// we will also load custom systems here in the future
+	loadCustomVirtualSystems();
+	// Now see which ones are enabled
+	loadEnabledListFromSettings();
+	// add to the main System Vector, and create Views as needed
+	updateSystemsList();
+}
+
+void VirtualSystemManager::updateSystemsList()
+{
+	// remove all Virtual Systems
+	for(auto sysIt = SystemData::sSystemVector.begin(); sysIt != SystemData::sSystemVector.end(); )
+	{
+		if ((*sysIt)->isVirtual())
+		{
+			sysIt = SystemData::sSystemVector.erase(sysIt);
+		}
+		else
+		{
+			sysIt++;
+		}
+	}
+
+	// add enabled ones
+	for(std::map<std::string, VirtualSystemData>::iterator it = mAllVirtualSystems.begin() ; it != mAllVirtualSystems.end() ; it++ )
+	{
+		if(it->second.isEnabled)
+		{
+			SystemData::sSystemVector.push_back(it->second.system);
+		}
+	}
+
+	// remove disabled gamelist views
+	// create gamelist views if needed
+	// iterate the map
+}
+
+void VirtualSystemManager::loadAutoVirtualSystems()
+{
+	for(std::map<std::string, VirtualSystemDecl>::iterator it = mVirtualSystemDecls.begin() ; it != mVirtualSystemDecls.end() ; it++ )
+	{
+		VirtualSystemDecl sysDecl = it->second;
+		if (!sysDecl.isCustom && !findVirtualSystem(sysDecl.name))
+		{
+			SystemData* newSys = new SystemData(sysDecl.name, sysDecl.longName, mVirtualEnvData, sysDecl.themeFolder, true);
+
+			FileData* rootFolder = newSys->getRootFolder();
+			FileFilterIndex* index = newSys->getIndex();
+			for(auto sysIt = SystemData::sSystemVector.begin(); sysIt != SystemData::sSystemVector.end(); sysIt++)
+			{
+				if ((*sysIt)->getName() != "retropie") {
+					std::vector<FileData*> files = (*sysIt)->getRootFolder()->getFilesRecursive(GAME);
+					for(auto gameIt = files.begin(); gameIt != files.end(); gameIt++)
+					{
+						bool include = (*gameIt)->getName() != "kodi";
+						switch(sysDecl.type) {
+							case AUTO_LAST_PLAYED:
+								include = include && (*gameIt)->metadata.get("playcount") > "0";
+								break;
+							case AUTO_FAVORITES:
+								include = include && (*gameIt)->metadata.get("favorite") == "true";
+								break;
+						}
+
+						if (include) {
+							FileData* newGame = new FileData(*gameIt, newSys);
```

Comment:
So this is essentially a duplication of something that is already known. How bad will this affect performance, especially in the case of AUTO_ALL_GAMES, for people with very large libraries?